### ISSUE TITLE
Add an option to automatically do save.image() / load('.RData')

### ISSUE THREAD

**Issue by flying-sheep** [IRkernel/IRkernel#175]
Things to consider:

When?
* Should it load before and save after each cell?
* Should it load when starting the kernel and save when stopping the kernel or closing the notebook?

Toggle?
* Should we deactivate it per default and add an option to activate it?
* Should we activate it per default unless a kernelspec option is set (since the feature has an effect before running a cell, we can’t just add an option to deactivate it)

Where?
* Should it load/save a `.RData`  next to the notebook or in the server root?

**Comment by vladchimescu** [comment]
The easiest solution would be to save before closing and load upon start as you suggest. This definitely has to be configurable, since as you mention .RData can be quite large for some Rnotebooks. 
Per default I would switch off autosave and autoload, since some users might note be aware of this feature (although when I started out with IRkernel I assumed that .RData is automatically saved and could be manually loaded).
Perhaps, the default option should be to save the .RData file in the directory of the notebook and give an option to override that and write to server root?
What about naming convention for associated .RData? Same as the notebook file name? A time stamp?

**Comment by flying-sheep** [comment]
> save before closing and load upon start as you suggest.

yes, if possible. maybe there is no hook for “browser tab closed”

> Perhaps, the default option should be to save the .RData file in the directory of the notebook and give an option to override that and write to server root?

hmm, not sure…

> What about naming convention for associated .RData? Same as the notebook file name? A time stamp?

`save.image()` and RStudio both use plain `.RData` (i.e. a hidden file), maybe that.

**Comment by poplav** [comment]
Perhaps the KISS solution is:

Have the kernel read an environment variable say ```r_kernel_work_space_file```.  If the environment variable exists on initialization of the kernel ```load(Sys.getenv("r_kernel_work_space_file"))```.  Then on a shutdown_request you would if environment variable exists ```save.image(file = Sys.getenv("r_kernel_work_space_file"))```

Then update the readme to say if one would like to specify a persistent workspace/data environment docker run with the environment variable r_kernel_work_space_file passed in/specified to the RData/workspace file.

**Comment by takluyver** [comment]
That sounds like it would get messy unless you're careful to only ever have
one R kernel running at a time.

**Comment by poplav** [comment]
Ah, true.  You would have each kernel instance/notebook spawned share that workspace file.  A regular user would definitely get confused when variables bleed between notebooks.  
The above idea though may be useful if you define your "global" workspace comprising of multiple notebooks in some enclosed/dockerized environment.  This would be an advanced case where they would really have to understand the implications.  
I don't think the kernel has the concept of a notebook id/info, so that route wouldn't work.  You would need some way of querying the client.  
Also, I imagine complications would arise if you wanted to save/load say SparkR variables.
